### PR TITLE
New content: IPLD + Filecoin section 

### LINKED ISSUE #943: Fill in IPLD section
Write the IPLD section. There is an outline here: https://github.com/filecoin-project/specs/blame/master/src/systems/filecoin_nodes/repository/ipldstore/_index.md#L19

Split out from #936

### PR DESCRIPTION
**Pull request by har00ga**
fixes #943 

This PR contains the initial draft for the IPLD section of the Filecoin specification, based on the outline provided in the above ticket and the doc produced after the conversation between @mishmosh @warpfork and myself. Definitely need more information on IPLDStores as of now, and still need to expand on how IPLD is actually used within Filecoin.

### CODE REVIEW

File: src/systems/filecoin_nodes/repository/ipldstore/_index.md [hunk 1]
```diff
@@ -5,25 +5,31 @@ title: "IpldStore - Local Storage for hash-linked data"
...
+IPLD is a set of libraries which allow for the interoperability of primitive hash-based data structures present in most distributed systems and protocols. It provides a fundamental 'common language' to cryptographic hashing, enabling data to be verifiably referenced and retrieved from other existing protocols. For example, a user can reference a git commit in a blockchain transaction to create an immutible copy and timestamp, or a data from a DHT can be refenced and linked to in a smart contract. <- assumption being made on last one -->
+
+<-- needs expansion of hash linked data, ipfs -->
+
+## IPLD in filecoin
+
+On the Filecoin network, all datastructures are stored in [IPLD](https://ipld.io) format, a data format akin to json built for storage, retrieval and traversal of hash-linked data DAGs. Files and data stored on the Filecoin network are not required to be in IPLD format, though it is useful, as IPLD allows the use of selectors[link] to retrieve a smaller subset of the total stored data, as opposed to inefficiently downloading the data set entirely.
```
Comment (mishmosh) [inline src/systems/filecoin_nodes/repository/ipldstore/_index.md:14]:
```suggestion
On the Filecoin network, all datastructures are stored in [IPLD](https://ipld.io) format, a data format akin to json built for storage, retrieval and traversal of hash-linked data DAGs. Files and data stored on the Filecoin network are not required to be in IPLD format, though it is useful, as IPLD allows the use of [selectors](https://github.com/ipld/specs/blob/master/selectors/selectors.md) to retrieve a smaller subset of the total stored data, as opposed to inefficiently downloading the data set entirely.
```

Comment (mishmosh) [reply src/systems/filecoin_nodes/repository/ipldstore/_index.md:14]: Make the 2 usages even more clear, eg with bullets.

File: src/systems/filecoin_nodes/repository/ipldstore/_index.md [hunk 2]
```diff
@@ -5,25 +5,33 @@ title: "IpldStore - Local Storage for hash-linked data"
 
 {{< readfile file="../../../../libraries/ipld/ipld.id" code="true" lang="go" >}}
 
-Filecoin datastructures are stored in [IPLD](https://ipld.io) format, a data format akin to json built for storage, retrieval and traversal of hash-linked data DAGs.
+IPLD is a set of libraries which allow for the interoperability of primitive hash-based data structures present in most distributed systems. It provides a fundamental 'common language' to cryptographic hashing, enabling data to be verifiably referenced and retrieved between two independent protocols. For example, a user can reference a git commit in a blockchain transaction to create an immutible copy and timestamp, or a data from a DHT can be refenced and linked to in a smart contract. <- assumption being made on last one -->
```
Comment (rvagg) [inline src/systems/filecoin_nodes/repository/ipldstore/_index.md:8]: it might be best to stick to "content addressed" rather than "hash-based"

**Comment by rvagg** [comment]
Have had a stab at this myself @ #956, use as you see fit, let me know if you need more. We'll hopefully have a nice graphic for you soon too.

File: src/systems/filecoin_nodes/repository/ipldstore/_index.md [hunk 3]
```diff
@@ -5,25 +5,38 @@ title: "IpldStore - Local Storage for hash-linked data"
...
+- Files and data stored on the Filecoin network may also be stored in IPLD format. While this is not required, it offers the benefit of supporting [selectors](https://github.com/ipld/specs/blob/master/selectors/selectors.md) to retrieve a smaller subset of the total stored data, as opposed to inefficiently downloading the data set entirely.
+
+IPLD handles marshalling and unmarshalling via a suite of codecs. IPLD codecs may support the complete Data Model or part of the Data Model. Two codecs that support the complete Data Model are [DAG-CBOR](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-cbor.md) and [DAG-JSON](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-json.md). 
+
+Filecoin uses the **DAG-CBOR** codec for the serialization and deserialization of it's data structures and interacts with that data using the IPLD Data Model, upon which various tools are built. IPLD [Paths](https://github.com/ipld/specs/blob/master/data-model-layer/paths.md) are also used to address specific nodes within a linked data structure.
```
Comment (mishmosh) [inline src/systems/filecoin_nodes/repository/ipldstore/_index.md:21]:
```suggestion
Filecoin uses the **DAG-CBOR** codec for the serialization and deserialization of its data structures and interacts with that data using the IPLD Data Model, upon which various tools are built. IPLD [Paths](https://github.com/ipld/specs/blob/master/data-model-layer/paths.md) are also used to address specific nodes within a linked data structure.
```

File: src/systems/filecoin_nodes/repository/ipldstore/_index.md [hunk 4]
```diff
@@ -5,25 +5,38 @@ title: "IpldStore - Local Storage for hash-linked data"
...
+For a more in-depth explanation of IPLD's role in Filecoin, see [this section] of the specification.
+
+## IPLD in filecoin
+
+On the Filecoin network, IPLD is used in two ways:
+
+- All system datastructures are stored in [IPLD](https://ipld.io) format, a data format akin to JSON but designed for storage, retrieval and traversal of hash-linked data DAGs.
+- Files and data stored on the Filecoin network may also be stored in IPLD format. While this is not required, it offers the benefit of supporting [selectors](https://github.com/ipld/specs/blob/master/selectors/selectors.md) to retrieve a smaller subset of the total stored data, as opposed to inefficiently downloading the data set entirely.
+
+IPLD handles marshalling and unmarshalling via a suite of codecs. IPLD codecs may support the complete Data Model or part of the Data Model. Two codecs that support the complete Data Model are [DAG-CBOR](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-cbor.md) and [DAG-JSON](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-json.md).
```
Comment (mishmosh) [inline src/systems/filecoin_nodes/repository/ipldstore/_index.md:19]: define marshalling & unmarshalling?

File: src/systems/filecoin_nodes/repository/ipldstore/_index.md [hunk 5]
```diff
@@ -5,25 +5,38 @@ title: "IpldStore - Local Storage for hash-linked data"
 
 {{< readfile file="../../../../libraries/ipld/ipld.id" code="true" lang="go" >}}
 
-Filecoin datastructures are stored in [IPLD](https://ipld.io) format, a data format akin to json built for storage, retrieval and traversal of hash-linked data DAGs.
+IPLD is a set of libraries which allow for the interoperability of content-addressed data structures present in most distributed systems. It provides a fundamental 'common language' to primitive cryptographic hashing, enabling data to be verifiably referenced and retrieved between two independent protocols. For example, a user can reference a git commit in a blockchain transaction to create an immutible copy and timestamp, or a data from a DHT can be refenced and linked to in a smart contract. 
+
+For a more in-depth explanation of IPLD's role in Filecoin, see [this section] of the specification.
```
Comment (mishmosh) [inline src/systems/filecoin_nodes/repository/ipldstore/_index.md:10]: "this section" - which sections do you plan to point to?

File: src/systems/filecoin_nodes/repository/ipldstore/_index.md [hunk 6]
```diff
@@ -5,25 +5,38 @@ title: "IpldStore - Local Storage for hash-linked data"
 
 {{< readfile file="../../../../libraries/ipld/ipld.id" code="true" lang="go" >}}
 
-Filecoin datastructures are stored in [IPLD](https://ipld.io) format, a data format akin to json built for storage, retrieval and traversal of hash-linked data DAGs.
+IPLD is a set of libraries which allow for the interoperability of content-addressed data structures present in most distributed systems. It provides a fundamental 'common language' to primitive cryptographic hashing, enabling data to be verifiably referenced and retrieved between two independent protocols. For example, a user can reference a git commit in a blockchain transaction to create an immutible copy and timestamp, or a data from a DHT can be refenced and linked to in a smart contract.
```
Comment (mishmosh) [inline src/systems/filecoin_nodes/repository/ipldstore/_index.md:8]:
```suggestion
IPLD is a set of libraries which allow for the interoperability of content-addressed data structures across different distributed systems. It provides a fundamental 'common language' to primitive cryptographic hashing, enabling data to be verifiably referenced and retrieved between two independent protocols. For example, a user can reference a git commit in a blockchain transaction to create an immutable copy and timestamp, or a data from a DHT can be referenced and linked to in a smart contract. 
```

File: src/systems/filecoin_nodes/repository/ipldstore/_index.md [hunk 7]
```diff
@@ -5,25 +5,38 @@ title: "IpldStore - Local Storage for hash-linked data"
...
 
 TODO:
 
-- What is IPLD
-  - hash linked data
-  - from IPFS
-- Why is it relevant to filecoin
-  - all network datastructures are definitively IPLD
-  - all local datastructures can be IPLD
 - What is an IpldStore
```
Comment (mishmosh) [inline src/systems/filecoin_nodes/repository/ipldstore/_index.md:40]: I think lines 40-42 can be deleted now that lines 29 onward exist?

File: src/systems/filecoin_nodes/repository/ipldstore/_index.md (hunk 4 above)
Comment (har00ga) [reply src/systems/filecoin_nodes/repository/ipldstore/_index.md:19]: I think that's something most developers will already understand, but may be useful to include in the glossary as well

File: src/systems/filecoin_nodes/repository/ipldstore/_index.md (hunk 5 above)
Comment (har00ga) [reply src/systems/filecoin_nodes/repository/ipldstore/_index.md:10]: The section that [this PR](https://github.com/filecoin-project/specs/pull/956) implements, wasn't sure what the link slug would look like ultimately though

Comment (mishmosh) [reply src/systems/filecoin_nodes/repository/ipldstore/_index.md:10]: Ah, let's remove the mention until that lands then.

**Comment by rvagg** [comment]
Fine for now, there's still a reference to "IPLD format ... similar to JSON" but then below clarifies that it's DAG-CBOR, so it's not crystal clear that IPLD _isn't_ a format, but an abstraction layer that incorporates many formats (that we call codecs).

When we get our diagram finished we can bring it in here and clean up some of the language. Will try and get that all done soon.

**Pull request merged**